platform/linux-generic/odp_schedule_basic.c

```diff
@@ -1423,21 +1458,29 @@ static inline int do_schedule(odp_queue_t *out_q, odp_event_t out_ev[], uint32_t
 	if (odp_unlikely(num_grp == 0))
 		return 0;
 
+	my_groups = sched_local.grp_mask;
 	first_id = sched_local.grp_idx;
 	sched_local.grp_idx = (first_id + 1) % num_grp;
 
 	for (prio = 0; prio < NUM_PRIO; prio++) {
 		grp_id = first_id;
 
+		if ((sched->prio_grp_mask[prio] & my_groups) == 0) {
```

Comment:
This is actually a data race as schedule_creater_queue() may just be updating the mask through prio_grp_mask_set(). I suppose it is intentional, the idea being that this is just an optimization and it does not matter if in some scheduling rounds the mask seen is wrong. In that case it would be good to have a comment about it. Even better would be to use an atomic type (and a relaxed read) for the mask to get rid of the data race that strictly speaking leads to undefined behaviour.

There is the same issue with the existing prio_q_mask field.

Reply:
Added a comment and wrapped into a function in v3.

```
		if (prio_grp_mask_check(prio, my_groups) == 0) {
			/* My groups do not have queues at this priority level, continue to
			 * the next level.
			 *
			 * As a performance optimization, prio_grp_mask[] is checked without
			 * taking the lock. Masks change infrequently and usage of an old mask
			 * just leads into searching events from old priority levels,
			 * new levels are likely used on the next schedule call. */
			continue;
		}
```

Reply:
That does not fix the problem that it is still a data race (even if a harmless one in practise). Dynamic analyzers, like thread-sanitizer, may complain about it. I am not proposing using a lock but just changing the type to suitable atomic type (which gets rid of the data race).

Reply:
Changed prio_grp_mask[] typep to atomic in v4.

Reply:
The change looks good (it is ok to leave prio_q_mask as it is for now since ODP does not have 8-bit atomics). The commit message could be improved since now there is no data race in prio_grp_mask anymore (atomic variables do not create data races). Maybe rephrase like this:

Use a relaxed atomic load in schedule loop to remove a data race
and to document that reading prio_grp_mark when it might be written
is intentional.